File: engine/src/flutter/lib/web_ui/skwasm/animated_image.cpp
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "export.h"
#include "live_objects.h"
#include "skwasm_support.h"

#include "third_party/skia/include/android/SkAnimatedImage.h"
#include "third_party/skia/include/codec/SkAndroidCodec.h"
#include "third_party/skia/include/codec/SkCodec.h"
#include "third_party/skia/include/codec/SkGifDecoder.h"
#include "third_party/skia/include/codec/SkWebpDecoder.h"

#include <memory>

namespace {
std::unique_ptr<SkCodec> getCodecForData(SkData* data) {
  if (SkGifDecoder::IsGif(data->data(), data->size())) {
    return SkGifDecoder::Decode(sk_ref_sp(data), nullptr);
  }
  if (SkWebpDecoder::IsWebp(data->data(), data->size())) {
    return SkWebpDecoder::Decode(sk_ref_sp(data), nullptr);
  }
  return nullptr;
}
}  // namespace

SKWASM_EXPORT SkAnimatedImage* animatedImage_create(SkData* data,
                                                    int targetWidth,
                                                    int targetHeight) {
  liveAnimatedImageCount++;
  auto codec = getCodecForData(data);
  if (!codec) {
    printf("Failed to create codec for animated image.\n");
    return nullptr;
  }

  auto aCodec = SkAndroidCodec::MakeFromCodec(std::move(codec));
  if (aCodec == nullptr) {
    printf("Failed to create codec for animated image.\n");
    return nullptr;
  }

  if (targetWidth == 0 || targetHeight == 0) {
    return SkAnimatedImage::Make(std::move(aCodec)).release();
  }

  return SkAnimatedImage::Make(
             std::move(aCodec),
             SkImageInfo::MakeUnknown(targetWidth, targetHeight),
             SkIRect::MakeWH(targetWidth, targetHeight), nullptr)
      .release();
}

SKWASM_EXPORT void animatedImage_dispose(SkAnimatedImage* image) {
  liveAnimatedImageCount--;
  image->unref();
}

SKWASM_EXPORT int animatedImage_getFrameCount(SkAnimatedImage* image) {
  return image->getFrameCount();
}

SKWASM_EXPORT int animatedImage_getRepetitionCount(SkAnimatedImage* image) {
  return image->getRepetitionCount();
}

SKWASM_EXPORT int animatedImage_getCurrentFrameDurationMilliseconds(
    SkAnimatedImage* image) {
  return image->currentFrameDuration();
}

SKWASM_EXPORT void animatedImage_decodeNextFrame(SkAnimatedImage* image) {
  image->decodeNextFrame();
}

SKWASM_EXPORT SkImage* animatedImage_getCurrentFrame(SkAnimatedImage* image) {
  liveImageCount++;
  return image->getCurrentFrame().release();
}

File: engine/src/flutter/lib/web_ui/lib/src/engine/canvaskit/animated_image.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

/// Uses image codecs supplied by the CanvasKit WASM bundle.
///
/// See also:
///
///  * `image_web_codecs.dart`, which uses the `ImageDecoder` supplied by the browser.
library animated_image;

import 'dart:async';
import 'dart:typed_data';

import 'package:ui/src/engine.dart';

/// The CanvasKit implementation of [BackendAnimatedImage].
///
/// This class acts as a simple, backend-specific adapter around CanvasKit's
/// native [SkAnimatedImage]. It delegates frame extraction and metadata queries
/// directly to the native object, and deletes the native WebAssembly object when
/// explicitly disposed.
class CkAnimatedImage implements BackendAnimatedImage {
  /// Decodes an animated image from a list of encoded bytes.
  CkAnimatedImage.decodeFromBytes(this._bytes, this.src, {this.targetWidth, this.targetHeight}) {
    skAnimatedImage = createSkAnimatedImage();
  }

  /// The underlying native Skia animated image object.
  late final SkAnimatedImage skAnimatedImage;
  final String src;
  final Uint8List _bytes;
  int _frameCount = 0;
  int _repetitionCount = -1;

  /// The target width requested by the user.
  ///
  /// Note: CanvasKit's WASM animated image decoder does not support native resizing
  /// during decode. The shared frontend (`_SkiaEngineCodec`) will apply a canvas-based
  /// scaling fallback if these dimensions are specified.
  final int? targetWidth;

  /// The target height requested by the user.
  final int? targetHeight;

  /// Invokes CanvasKit to instantiate the native decoder and read initial metadata.
  SkAnimatedImage createSkAnimatedImage() {
    final SkAnimatedImage? animatedImage = canvasKit.MakeAnimatedImageFromEncoded(_bytes);
    if (animatedImage == null) {
      throw ImageCodecException(
        'Failed to decode image data.\n'
        'Image source: $src',
      );
    }

    _frameCount = animatedImage.getFrameCount().toInt();
    _repetitionCount = animatedImage.getRepetitionCount().toInt();

    return animatedImage;
  }

  bool _disposed = false;
  bool get debugDisposed => _disposed;

  bool _debugCheckIsNotDisposed() {
    assert(!_disposed, 'This image has been disposed.');
    return true;
  }

  @override
  void dispose() {
    assert(!_disposed, 'Cannot dispose a codec that has already been disposed.');
    _disposed = true;
    skAnimatedImage.delete();
  }

  @override
  int get frameCount {
    assert(_debugCheckIsNotDisposed());
    return _frameCount;
  }

  @override
  int get repetitionCount {
    assert(_debugCheckIsNotDisposed());
    return _repetitionCount;
  }

  @override
  Future<BackendFrameInfo> getNextFrame() {
    assert(_debugCheckIsNotDisposed());

    // Query the display duration for the current frame.
    final int frameDurationMs = skAnimatedImage.currentFrameDuration().toInt();

    // Extract the current frame as a static SkImage and wrap it.
    final SkImage skImage = skAnimatedImage.makeImageAtCurrentFrame();

    final currentFrame = BackendFrameInfo(
      duration: Duration(milliseconds: frameDurationMs),
      image: CkImageDelegate(skImage),
    );

    // Advance the native decoder so the next frame is ready on the next call.
    skAnimatedImage.decodeNextFrame();

    return Future<BackendFrameInfo>.value(currentFrame);
  }
}
